sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,277 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(L, X, y, n_components, loss, outer, threshold=0.0):
+    n_samples, n_features = X.shape
+
+    L = L.reshape((n_components, n_features))
+
+    Lx = np.dot(X, L.T)  # n_samples x n_components
+    assert Lx.shape == (n_samples, n_components)
+
+    A = Lx[np.newaxis, :, :] - Lx[:, np.newaxis, :]  # n_samples x n_samples x n_components
+    assert A.shape == (n_samples, n_samples, n_components)
+
+    logp = -np.einsum("ijk,ijk->ij", A, A)  # n_samples x n_samples
+    np.fill_diagonal(logp, -np.inf)
+    logp -= sp.misc.logsumexp(logp, axis=1)[:, np.newaxis]
+    assert logp.shape == (n_samples, n_samples)
+
+    p = np.exp(logp)  # n_samples x n_samples
+    assert p.shape == (n_samples, n_samples)
+
+    class_neighbours = y[:, np.newaxis] == y[np.newaxis, :]
+
+    p_i = np.einsum("ij,ij->i", p, class_neighbours)
+
+    proba = np.copy(p)
+    if loss == 'l1':
+        proba *= p_i[:, np.newaxis]
+    mask = proba > threshold
+    proba *= mask
+
+    grad = np.einsum("ij,ijkl", proba, outer)
+    assert grad.shape == (n_features, n_features)
+
+    neighbours_proba = class_neighbours * p
+    if loss == 'kl':
+        nonzero = p_i > 1e-10
+        neighbours_proba[nonzero, :] /= p_i[nonzero, np.newaxis]
+
+    neighbours_mask = neighbours_proba > threshold
+    neighbours_proba *= neighbours_mask
+
+    neighbours_term = np.einsum("ij,ijkl", neighbours_proba, outer)
+
+    assert neighbours_term.shape == (n_features, n_features)
+    grad -= neighbours_term
+
+    function_value = 0
+    if loss == 'l1':
+        function_value = p_i.sum()
+    elif loss == 'kl':
+        function_value = np.log(p_i).sum()
+
+    grad = 2 * np.dot(L, grad)
+
+    return [-function_value, -grad.flatten()]
+
+
+def nca_semivectorized_oracle(L, X, y, n_components, loss, threshold=0.0):
+    n_samples, n_features = X.shape
+    L = L.reshape((n_components, n_features))
+
+    Lx = np.dot(X, L.T)  # n_samples x n_components
+    assert Lx.shape == (n_samples, n_components)
+
+    grad = np.zeros((n_features, n_features))
+    function_value = 0
+
+    distances = np.zeros((n_samples, n_samples))
+    for i in range(n_samples):
+        js = np.arange(i)
+        diff = Lx[i, :] - Lx[js, :]
```

Comment:
Yes, that's the same.

I did a quick test to compare this version to `euclidean_distances` (which I apparently forgot to push). Performance has improved, but not that significant, and even with this indexing my versions appears to be slower than  `euclidean_distances`. Thus I'm inclined to reuse that function rather than invent yet another bicycle.

Profiler data (maybe a bit useless, since it was different runs, and, probably, different initializations)
Manually calculating distances: https://gist.github.com/Barmaley-exe/799b8252edda3d933b91
`euclidean_distances`: https://gist.github.com/Barmaley-exe/6bafd7b2d34c37b9a424